Libcloud 574 routes and route tables

zerthimon opened the pull request:
Support for EC2 Route Tables and Routes.

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: EC2RouteTable class, ex_list_route_tables method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: EC2Route and EC2SubnetAssociation classes

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: docstrings

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: ex_create_route_table() method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: ex_delete_route_table() method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: ex_associate_route_table() method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: ex_dissociate_route_table() method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: revert accidently modified header

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: ex_replace_route_table_association() method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: ex_create_route() method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: ex_delete_route() method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: ex_replace_route() method

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: Kami's code review and some lint fixes

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: more lint fixes

zerthimon commented:
@Kami is there anything else I should do here ?

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -1730,6 +1739,105 @@ def __repr__(self):
+    :type       id: ``str``
+
+    :param      routes: A list of routes in the route table.
+    :type       routes: ``list`` of :class:`EC2Route`
```
Please add `.` before the classes which are located in the same module. In this case:

``` python
 :type       routes: ``list`` of :class:`.EC2Route`
```

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -1730,6 +1739,105 @@ def __repr__(self):
+        self.extra = extra or {}
+
+    def __repr__(self):
+        return (('<EC2RouteTable: id=%s') % (self.id))
```
Missing closing >

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -1730,6 +1739,105 @@ def __repr__(self):
+    """
+    Class which stores information about VPC Route Tables.
+
+    :param      id: The ID of the route table.
```
Please move docstring which documents the arguments inside the `__init__` method.

Same goes for other classes.

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -1730,6 +1739,105 @@ def __repr__(self):
+        self.vpc_peering_connection_id = vpc_peering_connection_id
+
+    def __repr__(self):
+        return (('<EC2Route: cidr=%s') % (self.cidr))
```
Same here, missing >

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -3927,6 +4035,311 @@ def ex_detach_internet_gateway(self, gateway, network):
 
         return element == 'true'
 
+    def ex_list_route_tables(self, rtable_ids=None, filters=None):
```
Probably would be better to be explicit and and consistent with the method name and call the argument `route_table_ids`

Kami commented:
Ok, reviewed the whole pull request.

There are some minor issues left, but it mostly looks good. Next step would be adding tests for this new functionality. Let us know if you need help with that.

Kami commented:
@zerthimon Let me know once those issues have been addressed and I will get it merged into trunk.

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: Review fixes, rewrite methods to work with objects instead of ID strings

zerthimon committed:
LIBCLOUD-574 Routes-and-Route-Tables: fix a long line

zerthimon commented:
@Kami 

Fixed issues that you found during code review.
Also changed ex_create_route() and ex_replace_route() to accept objects and extract IDs, instead of passing IDs as strings.

Kami commented:
Merged into trunk. Thanks.
